### PR TITLE
Add support for overlay volume mounts in podman.

### PR DESCRIPTION
**Pull request by QiWang19**
Add support for overlay volume mounts in podman.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>
Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

File: libpod/container.go [hunk 1]
```diff
@@ -447,6 +449,19 @@ type ContainerNamedVolume struct {
 	Options []string `json:"options,omitempty"`
 }
 
+// ContainerOverlayVolume is a overlay volume that will be mounted into the
+// container. Each volume is a libpod Volume present in the state.
+type ContainerOverlayVolume struct {
+	// Destination is the absolute path where the mount will be placed in the container.
+	Dest string `json:"dest"`
+	// Type specifies the mount kind.
+	Type string `json:"type,omitempty" platform:"linux,solaris"`
```
Comment (mheon) [inline libpod/container.go:458]: Why is this here? I think we know type based on the fact that it's an overlay mount?

File: libpod/container.go [hunk 2]
```diff
@@ -447,6 +449,19 @@ type ContainerNamedVolume struct {
...
+	// Source specifies the source path of the mount.
+	Source string `json:"source,omitempty"`
```
Comment (mheon) [inline libpod/container.go:460]: Can't we have multiple sources in an overlay mount?

File: libpod/container_internal.go [hunk 3]
```diff
@@ -1425,7 +1425,7 @@ func (c *Container) mountStorage() (_ string, deferredErr error) {
 		defer func() {
 			if deferredErr != nil {
 				if err := c.unmountSHM(c.config.ShmDir); err != nil {
-					logrus.Errorf("Error unmounting SHM for container %s after mount error: %v", c.ID(), err)
+					logrus.Errorf(err.Error())
```
Comment (mheon) [inline libpod/container_internal.go:1428]: Why did this change?

File: libpod/container_internal.go [hunk 4]
```diff
@@ -1582,11 +1582,16 @@ func (c *Container) cleanupStorage() error {
 
 	for _, containerMount := range c.config.Mounts {
 		if err := c.unmountSHM(containerMount); err != nil {
-			if cleanupErr != nil {
-				logrus.Errorf("Error unmounting container %s: %v", c.ID(), cleanupErr)
-			}
+			logrus.Errorf(err.Error())
```
Comment (mheon) [inline libpod/container_internal.go:1585]: Again, I don't think this should have changed

File: libpod/container_internal_linux.go [hunk 5]
```diff
@@ -56,7 +57,9 @@ func (c *Container) mountSHM(shmOptions string) error {
 func (c *Container) unmountSHM(mount string) error {
 	if err := unix.Unmount(mount, 0); err != nil {
 		if err != syscall.EINVAL && err != syscall.ENOENT {
-			return errors.Wrapf(err, "error unmounting container %s SHM mount %s", c.ID(), mount)
+			logrus.Warnf("container %s failed to unmount %s : %v", c.ID(), mount, err)
```
Comment (mheon) [inline libpod/container_internal_linux.go:60]: I think this should stay an error

File: libpod/container_internal_linux.go [hunk 6]
```diff
@@ -56,7 +57,9 @@ func (c *Container) mountSHM(shmOptions string) error {
...
+		} else {
+			logrus.Debugf("container %s failed to unmount %s : %v", c.ID(), mount, err)
```
Comment (mheon) [inline libpod/container_internal_linux.go:62]: This is a duplicate of what was printed below

File: libpod/container_internal_linux.go [hunk 7]
```diff
@@ -316,6 +319,19 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
 		}
 	}
 
+	// Add overlay volumes
+	for _, overlayVol := range c.config.OverlayVolumes {
+		contentDir, err := overlay.TempDir(c.config.StaticDir, c.RootUID(), c.RootGID())
```
Comment (mheon) [inline libpod/container_internal_linux.go:324]: Should this be the root UID/GID or the UID/GID of the user running in the container?

File: libpod/container_internal_linux.go [hunk 8]
```diff
@@ -316,6 +319,19 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
...
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create TempDir in the %s directory", c.config.StaticDir)
+		}
+		overlayMount, err := overlay.Mount(contentDir, overlayVol.Source, overlayVol.Dest, c.RootUID(), c.RootGID(), c.runtime.store.GraphOptions())
```
Comment (mheon) [inline libpod/container_internal_linux.go:328]: It looks like this discards the options from the overlayVol?

File: libpod/options.go [hunk 9]
```diff
@@ -1394,6 +1394,41 @@ func WithNamedVolumes(volumes []*ContainerNamedVolume) CtrCreateOption {
 	}
 }
 
+// WithOverlayVolumes adds the given overlay volumes to the container.
+func WithOverlayVolumes(volumes []*ContainerOverlayVolume) CtrCreateOption {
+	return func(ctr *Container) error {
```
Comment (mheon) [inline libpod/options.go:1399]: All of this should be done in Specgen - `WithOverlayVolumes` should just populate `c.config.OverlayVolumes`

File: pkg/specgen/specgen.go [hunk 10]
```diff
@@ -432,6 +434,10 @@ type NamedVolume struct {
 	Options []string
 }
 
+// OverlayVolume holds information about a overlay volume that will be mounted into
+// the container.
+type OverlayVolume spec.Mount
```
Comment (mheon) [inline pkg/specgen/specgen.go:439]: This doesn't make sense, a lot of fields in spec.Mount are not used.

File: pkg/spec/storage.go [hunk 11]
```diff
@@ -445,7 +445,7 @@ func getBindMount(args []string) (spec.Mount, error) {
 			}
 			setExec = true
 			newMount.Options = append(newMount.Options, kv[0])
-		case "shared", "rshared", "private", "rprivate", "slave", "rslave", "Z", "z":
+		case "shared", "rshared", "private", "rprivate", "slave", "rslave", "Z", "z", "O":
```
Comment (mheon) [inline pkg/spec/storage.go:448]: This shouldn't be necessary since we aren't handling `:O` as part of a regular volume

File: pkg/specgen/specgen.go [hunk 12]
```diff
@@ -190,6 +190,8 @@ type ContainerStorageConfig struct {
 	// there are conflicts.
 	// Optional.
 	Volumes []*NamedVolume `json:"volumes,omitempty"`
+	// Overlay volumes are named volumes that will be added to the container.
```
Comment (mheon) [inline pkg/specgen/specgen.go:193]: Please note that this is Optional.

File: cmd/podman/common/volumes.go [hunk 13]
```diff
@@ -104,12 +104,12 @@ func parseVolumes(volumeFlag, mountFlag, tmpfsFlag []string, addReadOnlyTmpfs bo
 	// Check for conflicts between named volumes and mounts
```
Comment (mheon) [inline cmd/podman/common/volumes.go:104]: We need conflict checks for the new overlay volumes, as we have for named volumes and normal bind mounts right now.

File: cmd/podman/common/volumes.go [hunk 14]
```diff
@@ -496,34 +501,51 @@ func getVolumeMounts(volumeFlag []string) (map[string]spec.Mount, map[string]*sp
...
 			// This is not a named volume
-			newMount := spec.Mount{
-				Destination: cleanDest,
-				Type:        string(TypeBind),
-				Source:      src,
-				Options:     options,
-			}
-			if _, ok := mounts[newMount.Destination]; ok {
-				return nil, nil, errors.Wrapf(errDuplicateDest, newMount.Destination)
+			for _, o := range options {
```
Comment (mheon) [inline cmd/podman/common/volumes.go:522]: This loop makes no sense. We should check every option to determine if `O` is present, and then add *after* the loop based on whether it is or is not.

File: libpod/container_internal_linux.go (hunk 8 above)
Comment (QiWang19) [reply libpod/container_internal_linux.go:328]: the overlay pkg has hardcoded options, @rhatdan can other options be used here?

Comment (rhatdan) [reply libpod/container_internal_linux.go:328]: Theoretically yes, they could be added. I guess you could specify RO, but a readonly overlay makes little sense.
The private, shared, slave might be able to be passed down.
For now we could throw an error when they are passed in, and wait for someone to complain.  I am not sure how well these options would work.

File: libpod/options.go (hunk 9 above)
Comment (QiWang19) [reply libpod/options.go:1399]: Do you mean `cmd/podman/common/specgen.go`?  Why can't keep this like `WithNamedVolumes` above?

Comment (mheon) [reply libpod/options.go:1399]: Hmm. Didn't realize that we checked again there.

We should retain the ProcessOptions part, but move the destination check to container_validate.go and have it check both named volumes and overlay volumes.

File: libpod/container_internal_linux.go (hunk 8 above)
Comment (mheon) [reply libpod/container_internal_linux.go:328]: I'd be fine with not allowing options at all here - just remove them from the struct, and error at the command line when someone tries to pass them.

Comment (mheon) [reply libpod/container_internal_linux.go:328]: (We can always re-add them later)

File: libpod/container.go (hunk 2 above)
Comment (QiWang19) [reply libpod/container.go:458]: Can multiple sources be used by allowing multiple `-v source:dest` with same dest?

Comment (mheon) [reply libpod/container.go:458]: I remember we had a discussion about this. @rhatdan Did we decide to allow multiple sources? If there's any possibility, we should make this an array.

Comment (rhatdan) [reply libpod/container.go:458]: If we want that, we probably need a different command, maybe something with --mount.  I think for --volume we should keep it simple and not complicate the option parsing to much.

Comment (QiWang19) [reply libpod/container.go:458]: Change back to treat multiple overlay sources with the same dest as conflict

**Comment by QiWang19** [comment]
@mheon PTAL

File: cmd/podman/common/volumes.go [hunk 15]
```diff
@@ -101,15 +101,21 @@ func parseVolumes(volumeFlag, mountFlag, tmpfsFlag []string, addReadOnlyTmpfs bo
...
 	for dest := range unifiedMounts {
 		if _, ok := unifiedVolumes[dest]; ok {
-			return nil, nil, errors.Wrapf(errDuplicateDest, "conflict at mount destination %v", dest)
+			return nil, nil, nil, errors.Wrapf(errDuplicateDest, "conflict at mount destination %v", dest)
+		}
+		if _, ok := overlayVolumes[dest]; ok {
+			return nil, nil, nil, errors.Wrapf(errDuplicateDest, "conflict at mount destination %v", dest)
 		}
 	}
 	for dest := range unifiedVolumes {
```
Comment (mheon) [inline cmd/podman/common/volumes.go:113]: We need another one of these for overlay volumes, to check that overlay volume destinations are not defined for named volumes or mounts.

File: cmd/podman/common/volumes.go [hunk 16]
```diff
@@ -496,34 +507,55 @@ func getVolumeMounts(volumeFlag []string) (map[string]spec.Mount, map[string]*sp
...
+				}
 			}
-			if _, ok := mounts[newMount.Destination]; ok {
-				return nil, nil, errors.Wrapf(errDuplicateDest, newMount.Destination)
+			if overlayFlag {
+				// This is a overlay volume
+				newOverlayVol := new(specgen.OverlayVolume)
+				newOverlayVol.Destination = cleanDest
+				newOverlayVol.Source = src
+				newOverlayVol.Options = options
```
Comment (mheon) [inline cmd/podman/common/volumes.go:542]: Why are we setting this if we explicitly require no options above?

File: cmd/podman/common/volumes.go [hunk 17]
```diff
@@ -496,34 +507,55 @@ func getVolumeMounts(volumeFlag []string) (map[string]spec.Mount, map[string]*sp
...
+				if _, ok := overlayVolumes[newOverlayVol.Destination]; ok {
```
Comment (mheon) [inline cmd/podman/common/volumes.go:543]: Why are we doing this here? This should be done as part of `parseVolumes` where the other conflict checks are done.

Comment (mheon) [reply cmd/podman/common/volumes.go:543]: Ah, nevermind, this is for internal conflicts.

File: docs/source/markdown/podman-create.1.md [hunk 18]
```diff
@@ -930,6 +939,37 @@ content label. Shared volume labels allow all containers to read/write content.
...
+
+  Subsequent executions of the container will see the original source directory
+content, any changes from previous container executions no longer exists.
+
+  One use case of the overlay mount is sharing the package cache from the
+host into the container to allow speeding up builds.
+
+  Note:
+
+     - The `O` flag conflicts with the `Z` and `z` flags.
```
Comment (mheon) [inline docs/source/markdown/podman-create.1.md:960]: It conflicts with all options now, doesn't it?

File: libpod/container.go [hunk 19]
```diff
@@ -447,6 +449,17 @@ type ContainerNamedVolume struct {
...
 
+// ContainerOverlayVolume is a overlay volume that will be mounted into the
+// container. Each volume is a libpod Volume present in the state.
+type ContainerOverlayVolume struct {
+	// Destination is the absolute path where the mount will be placed in the container.
+	Dest string `json:"dest"`
+	// Source specifies the source path of the mount.
+	Source string `json:"source,omitempty"`
+	// Options are fstab style mount options.
+	Options []string `json:"options,omitempty"`
```
Comment (mheon) [inline libpod/container.go:460]: Didn't we remove volumes above?

Comment (mheon) [reply libpod/container.go:460]: Sorry, remove options* above

File: pkg/util/mountOpts.go [hunk 20]
```diff
@@ -100,12 +100,12 @@ func ProcessOptions(options []string, isTmpfs bool, sourcePath string) ([]string
 				return nil, errors.Wrapf(ErrDupeMntOption, "only one of 'rbind' and 'bind' can be used")
 			}
 			foundBind = true
-		case "z", "Z":
+		case "z", "Z", "O":
```
Comment (mheon) [inline pkg/util/mountOpts.go:103]: We should probably enforce that O is only allowed if it's the only option here?

File: cmd/podman/common/volumes.go (hunk 16 above)
Comment (QiWang19) [reply cmd/podman/common/volumes.go:542]: Do you mean the above checking len(options)? the above is not requring no options, it's requiring that only one option is allowed.

File: pkg/util/mountOpts.go (hunk 20 above)
Comment (QiWang19) [reply pkg/util/mountOpts.go]: Added a case to handle 'O', but can we leave the `ProcessOptions` as it is without checking the 'O' option, since the error threw earlier when paring the CLI?

File: libpod/container.go (hunk 19 above)
Comment (QiWang19) [reply libpod/container.go:460]: Are you suggesting changing the comment and variable name to option, or removing the Options field?

File: cmd/podman/common/volumes.go (hunk 16 above)
Comment (mheon) [reply cmd/podman/common/volumes.go:542]: Yes - and that option is `O` which we should be discarding. no options that are not O are permissable, so Overlay Volumes don't use options right now.

**Comment by QiWang19** [comment]
@mheon PTAL

**Comment by QiWang19** [comment]
@mheon @rhatdan PTAL

File: test/e2e/run_volume_test.go [hunk 21]
```diff
@@ -450,4 +451,63 @@ VOLUME /test/`
 		Expect(data[0].Mounts[0].Source).To(Equal("/tmp"))
 		Expect(data[0].Mounts[0].Destination).To(Equal("/test"))
 	})
+
+	It("podman run with overlay volume flag", func() {
```
Comment (mheon) [inline test/e2e/run_volume_test.go:455]: Can we get some tests on conflicts? Adding an overlay volume at the same path as a named volume or bind mount or tmpfs should error.

File: libpod/container_internal_linux.go [hunk 22]
```diff
@@ -319,6 +320,29 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
...
 	}
 
+	// Add overlay volumes
+	uid := c.RootUID()
+	gid := c.RootGID()
+	idMappings, err := c.IDMappings()
+	if err != nil {
+		return nil, err
+	}
+	if len(idMappings.UIDMap) > 0 && len(idMappings.GIDMap) > 0 {
```
Comment (mheon) [inline libpod/container_internal_linux.go:330]: Do we need this? I thought RootUID and RootGID already looked through the mappings?

**Comment by mheon** [comment]
One more comment, then LGTM

@rhatdan @vrothberg @giuseppe @baude PTAL

File: libpod/container_internal_linux.go (hunk 22 above)
Comment (QiWang19) [reply libpod/container_internal_linux.go:330]: Add this because https://github.com/containers/podman/pull/6895#discussion_r451133191, is this the right way to get the uid/gid of the user running in the container?

Comment (rhatdan) [reply libpod/container_internal_linux.go:330]: @giuseppe ^^

**Comment by rhatdan** [comment]
LGTM

File: libpod/container_internal_linux.go (hunk 22 above)
Comment (mheon) [reply libpod/container_internal_linux.go:330]: I'd like this answered before merge, otherwise LGTM

Comment (QiWang19) [reply libpod/container_internal_linux.go:330]: @giuseppe PTAL

**Comment by rhatdan** [comment]
@giuseppe We are waiting for your review so we can get this merged.

File: libpod/container_internal_linux.go (hunk 22 above)
Comment (giuseppe) [reply libpod/container_internal_linux.go:330]: yes, I don't think we need the extra check, RootUID/RootGID are enough.

Also it seems to not check the container id but just get the first id available, it can be different than root as the mappings are not ordered

**Comment by giuseppe** [comment]
> @giuseppe We are waiting for your review so we can get this merged.

reviewed and left a comment

**Comment by mheon** [comment]
LGTM

**Comment by QiWang19** [comment]
@mheon @rhatdan PTAL

**Comment by rhatdan** [comment]
Great work @QiWang19 getting this merged.
